### PR TITLE
 Add implicits to provide delta(path) methods in Spark's read/write APIs for Scala

### PR DESCRIPTION
**Pull request by mahmoudmahdi24**
The main goal of this Pull Request is to create a ```delta``` function which replaces the ```.format("delta")``` we use whenever we want to read/write from/in a delta table .
I used the Scala "Pimp My Library" Pattern in order to extend the following APIs : 

- DataFrameReader
- DataStreamReader
- DataFrameWriter
- DataStreamWriter

I tested this feature by replacing, in the ```DeltaSourceSuite.scala```, all the ```.format("delta")``` by the function ```.delta(path)```

### CODE REVIEW

File: src/test/scala/org/apache/spark/sql/delta/DeltaSourceSuite.scala [hunk 1]
```diff
@@ -53,8 +53,7 @@ class DeltaSourceSuite extends DeltaSourceSuiteBase {
       new File(inputDir, "_delta_log").mkdir()
       val e = intercept[AnalysisException] {
         spark.readStream
-          .format("delta")
-          .load(inputDir.getCanonicalPath)
+          .delta(inputDir.getCanonicalPath)
```
Comment (jose-torres) [inline src/test/scala/org/apache/spark/sql/delta/DeltaSourceSuite.scala:56]: I don't know whether we need to refactor all the tests here. It seems better to just have a single new one to exercise each of the new implicits.

Comment (mahmoudmahdi24) [reply src/test/scala/org/apache/spark/sql/delta/DeltaSourceSuite.scala]: Thanks a lot for your feedback @jose-torres 🙏, I made the requested changes.

**Comment by mahmoudmahdi24** [comment]
Hello @jose-torres , @rahulsmahadev , Can you please approve this Pull Request if it's still relevant for you ? Thanks a lot

**Comment by jose-torres** [comment]
Hi! Very sorry, this slipped out of my inbox.

I'd suggest a few documentation changes:

* Avoid redundancy with the class/method signature in the doc.
* There's no need to use @param and @return in scenarios like this, where it's obvious from the signature what the arguments mean and no further context is being provided.
* Temporal indicators such as "now" aren't very future-proof - people in 6 months will be wondering when "now" is. Better to say something short and simple like "Usage: df.write.delta("path")".

I've run into this as a good general guide for documentation, although we don't have an official documentation guide for the project yet. https://www.monperrus.net/martin/how-to-write-good-API-documentation

**Comment by mahmoudmahdi24** [comment]
Hello @jose-torres , Thanks for the constructive feedback ! The article was quite interesting 🙏 
I took into account your feedback by removing redundant phrases and unuseful javadocs, and kept only the necessary ones.

**Comment by mahmoudmahdi24** [comment]
I have also rebased the branch and solved the conflicts with master.

**Comment by mahmoudmahdi24** [comment]
Hello @tdas @zsxwing , Can you please approve this Pull Request ?

**Comment by mahmoudmahdi24** [comment]
Hello @tdas @jose-torres , sorry for asking you again to take a look, can you please approve this PR ? I made the requested changes.
I think this modification can be helpful, it's great to be able to just read/write from/to a delta table without specifying the ```.format("delta")``` everytime.

File: src/main/scala/org/apache/spark/sql/delta/sources/DeltaDataSource.scala [hunk 2]
```diff
@@ -342,4 +343,44 @@ object DeltaDataSource extends DatabricksLogging {
       None
     }
   }
+
+  /**
+   * Extends the DataFrameReader API by adding a delta function
+   * Usage: spark.read.delta(path)
```
Comment (jaceklaskowski) [inline src/main/scala/org/apache/spark/sql/delta/sources/DeltaDataSource.scala:349]: Why not to use something like that would get formatted properly in scala/javadoc?

```
 * {{{
 * spark.read.delta(path)
 * }}}
```

File: src/main/scala/org/apache/spark/sql/delta/sources/DeltaDataSource.scala [hunk 3]
```diff
@@ -342,4 +343,44 @@ object DeltaDataSource extends DatabricksLogging {
...
+   */
+  implicit class DeltaDataFrameReader(val dfReader: DataFrameReader) extends AnyVal {
```
Comment (jaceklaskowski) [inline src/main/scala/org/apache/spark/sql/delta/sources/DeltaDataSource.scala:351]: nit: `reader` should just be fine

File: src/test/scala/org/apache/spark/sql/delta/DeltaSourceSuite.scala [hunk 4]
```diff
@@ -964,6 +964,71 @@ class DeltaSourceSuite extends DeltaSourceSuiteBase {
     }
   }
 
+  test("Read data using the .delta() method extended from the DataFrameReader API") {
+    withTempDir { tempDir =>
+      Seq(1, 2, 3).toDF("x").write.format("delta").save(tempDir.getCanonicalPath)
```
Comment (jaceklaskowski) [inline src/test/scala/org/apache/spark/sql/delta/DeltaSourceSuite.scala:969]: Does it really matter to use `"x"` for the columns? Clean it up for future readers.

File: src/test/scala/org/apache/spark/sql/delta/DeltaSourceSuite.scala [hunk 5]
```diff
@@ -964,6 +964,71 @@ class DeltaSourceSuite extends DeltaSourceSuiteBase {
...
+  }
+
+  test("read stream using the .delta() method extended from the DataStreamReader API") {
+    withTempDir { inputDir =>
+      val deltaLog = DeltaLog.forTable(spark, new Path(inputDir.toURI))
+      withMetadata(deltaLog, StructType.fromDDL("value STRING"))
+
+      val df = spark.readStream
+        .delta(inputDir.getCanonicalPath)
+        .filter($"value" contains "keep")
```
Comment (jaceklaskowski) [inline src/test/scala/org/apache/spark/sql/delta/DeltaSourceSuite.scala:994]: This `filter` is not needed here, is it?

**Comment by zsxwing** [comment]
@mahmoudmahdi24 could you update your PR to resolve the conflicts and address comments? Looks it's almost ready to merge.

**Comment by mahmoudmahdi24** [comment]
hello @zsxwing , I'll finish this today and let you know once it's ready.

File: src/test/scala/org/apache/spark/sql/delta/DeltaSourceSuite.scala (hunk 5 above)
Comment (mahmoudmahdi24) [reply src/test/scala/org/apache/spark/sql/delta/DeltaSourceSuite.scala]: I have removed it. Thanks for your feedback @jaceklaskowski

File: src/test/scala/org/apache/spark/sql/delta/DeltaSourceSuite.scala (hunk 4 above)
Comment (mahmoudmahdi24) [reply src/test/scala/org/apache/spark/sql/delta/DeltaSourceSuite.scala]: removed the column name as requested 👍

File: src/main/scala/org/apache/spark/sql/delta/sources/DeltaDataSource.scala (hunk 3 above)
Comment (mahmoudmahdi24) [reply src/main/scala/org/apache/spark/sql/delta/sources/DeltaDataSource.scala]: done 👍

File: src/main/scala/org/apache/spark/sql/delta/sources/DeltaDataSource.scala (hunk 2 above)
Comment (mahmoudmahdi24) [reply src/main/scala/org/apache/spark/sql/delta/sources/DeltaDataSource.scala]: done 👍

**Comment by mahmoudmahdi24** [comment]
Hello @zsxwing 👋 I have taken into account the PR's comments and resolved the conflicts. looking forward to have this feature merged. Thanks

File: src/main/scala/org/apache/spark/sql/delta/sources/DeltaDataSource.scala [hunk 6]
```diff
@@ -330,4 +331,56 @@ object DeltaDataSource extends DatabricksLogging {
...
   }
+
+  /**
+   * Extends the DataFrameReader API by adding a delta function
+   * Usage:
+   * {{{
+   * spark.read.delta(path)
+   * }}}
+   */
+  implicit class DeltaDataFrameReader(val reader: DataFrameReader) extends AnyVal {
```
Comment (zsxwing) [inline src/main/scala/org/apache/spark/sql/delta/sources/DeltaDataSource.scala:342]: DeltaDataSource is a private API. Could you create a new package file under `io.delta` and move these codes there? You can take a look at https://github.com/databricks/spark-avro/blob/branch-4.0/src/main/scala/com/databricks/spark/avro/package.scala as an example.

Comment (zsxwing) [reply src/main/scala/org/apache/spark/sql/delta/sources/DeltaDataSource.scala:342]: In addition,  the class name look not consistent. Can we use `DeltaDataFrameReader`, `DeltaDataFrameWriter`, `DeltaDataStreamReader`, `DeltaDataStreamWriter`?

File: src/main/scala/org/apache/spark/sql/delta/sources/DeltaDataSource.scala [hunk 7]
```diff
@@ -330,4 +331,56 @@ object DeltaDataSource extends DatabricksLogging {
...
+    def delta(path: String): DataFrame = {
+      reader.format("delta").load(path)
+    }
+  }
+
+  /**
+   * Extends the DataFrameReader API by adding a delta function
+   * Usage: spark.readStream.delta(path)
```
Comment (zsxwing) [inline src/main/scala/org/apache/spark/sql/delta/sources/DeltaDataSource.scala:350]:
```suggestion
   * Usage:
```

File: src/main/scala/org/apache/spark/sql/delta/sources/DeltaDataSource.scala [hunk 8]
```diff
@@ -330,4 +331,56 @@ object DeltaDataSource extends DatabricksLogging {
...
+   * }}}
+   */
+  implicit class DeltaDataFrameReader(val reader: DataFrameReader) extends AnyVal {
+    def delta(path: String): DataFrame = {
+      reader.format("delta").load(path)
+    }
+  }
+
+  /**
+   * Extends the DataFrameReader API by adding a delta function
```
Comment (zsxwing) [inline src/main/scala/org/apache/spark/sql/delta/sources/DeltaDataSource.scala:349]:
```suggestion
   * Extends the DataStreamReader API by adding a delta function
```

File: src/test/scala/org/apache/spark/sql/delta/DeltaSourceSuite.scala [hunk 9]
```diff
@@ -978,6 +976,70 @@ class DeltaSourceSuite extends DeltaSourceSuiteBase with DeltaSQLCommandTest {
...
+        AssertOnQuery { q => q.processAllAvailable(); true },
+        CheckAnswer("item1", "item2", "item3", "item4"),
+        AddToReservoir(inputDir, Seq("item5").toDF),
+        AssertOnQuery { q => q.processAllAvailable(); true },
+        CheckAnswer("item1", "item2", "item3", "item4", "item5")
+      )
+    }
+  }
+
+  test("write stream using the .delta() method extended from DataStreamWriter API") {
```
Comment (zsxwing) [inline src/test/scala/org/apache/spark/sql/delta/DeltaSourceSuite.scala:1023]: I think we can update this test to cover all four new APIs and remove the above three tests. For example,

```
Seq(1, 2, 3).toDF().write.delta(inputDir.toString)

val df = spark.readStream.delta(inputDir.toString)
val stream = df.writeStream
  .option("checkpointLocation", checkpointDir.toString)
  .delta(outputDir.toString)
stream.processAllAvailable()
stream.stop()

checkAnswer(spark.read.delta(outputDir.toString), Seq(Row(1), Row(2), Row(3)))
```

File: src/main/scala/org/apache/spark/sql/delta/sources/DeltaDataSource.scala (hunk 6 above)
Comment (mahmoudmahdi24) [reply src/main/scala/org/apache/spark/sql/delta/sources/DeltaDataSource.scala]: Great Idea 💯  I have taken into account your feedback and have created the ```io.delta.sources``` package.
Let me know if you want me to change the name.
I have also renamed the different implicit classes as requested.

File: src/main/scala/org/apache/spark/sql/delta/sources/DeltaDataSource.scala (hunk 7 above)
Comment (mahmoudmahdi24) [reply src/main/scala/org/apache/spark/sql/delta/sources/DeltaDataSource.scala]: Done 👍

File: src/main/scala/org/apache/spark/sql/delta/sources/DeltaDataSource.scala (hunk 8 above)
Comment (mahmoudmahdi24) [reply src/main/scala/org/apache/spark/sql/delta/sources/DeltaDataSource.scala]: Done 👍  It was an oversight on my part

File: src/test/scala/org/apache/spark/sql/delta/DeltaSourceSuite.scala (hunk 9 above)
Comment (mahmoudmahdi24) [reply src/test/scala/org/apache/spark/sql/delta/DeltaSourceSuite.scala]: Done 👍

**Comment by mahmoudmahdi24** [comment]
@zsxwing Thanks a lot for your feedback 🙏  I have taken into account your comments.

File: src/main/scala/io/delta/sources/package.scala [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.delta
+
+import org.apache.spark.sql.{DataFrame, DataFrameReader, DataFrameWriter}
+import org.apache.spark.sql.streaming.{DataStreamReader, DataStreamWriter, StreamingQuery}
+
+package object sources {
```
Comment (zsxwing) [inline src/main/scala/io/delta/sources/package.scala:22]: In a second thought, I think `import io.delta.implicits._` as this is similar to `import spark.implicits._` which people are familiar.

Could you add a `package object implicits` in the `src/main/scala/io/delta/package.scala` file?

**Review by zsxwing — CHANGES_REQUESTED** [review]
Some several minor comments. Otherwise LGTM.

File: src/test/scala/io/delta/sources/DeltaSourcesSuite.scala [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+    withTempDir { file1 =>
+      withTempDir { file2 =>
+        withTempDir { file3 =>
+          f(file1, file2, file3)
+        }
+      }
+    }
+  }
+
+  test("make sure that the delta sources works fine") {
```
[the remark this answers is not in the document]
Comment (mahmoudmahdi24) [reply src/test/scala/io/delta/sources/DeltaSourcesSuite.scala]: done 👍

File: src/main/scala/org/apache/spark/sql/delta/sources/DeltaDataSource.scala [hunk 12]
```diff
@@ -41,6 +41,7 @@ import org.apache.spark.sql.sources._
 import org.apache.spark.sql.streaming.OutputMode
 import org.apache.spark.sql.types.StructType
 import org.apache.spark.sql.util.CaseInsensitiveStringMap
+import org.apache.spark.sql.streaming.{DataStreamReader, DataStreamWriter, StreamingQuery}
```
[the remark this answers is not in the document]
Comment (mahmoudmahdi24) [reply src/main/scala/org/apache/spark/sql/delta/sources/DeltaDataSource.scala]: remove 👍

File: src/main/scala/io/delta/sources/package.scala (hunk 10 above)
Comment (mahmoudmahdi24) [reply src/main/scala/io/delta/sources/package.scala]: I think it's a good idea. I have renamed the package to implicits.

**Comment by mahmoudmahdi24** [comment]
I have taken into account your latest feedback @zsxwing . Thanks a lot 🙏

**Review by zsxwing — APPROVED** [review]

**Pull request closed without merging**